helpers/getLocalizeResources.js

```diff
@@ -0,0 +1,355 @@
+
+let cache = undefined;
+let cachePromise = undefined;
+let documentLocaleSettings = undefined;
+let queue = [];
+let state = StateIdle;
+let timer = 0;
+
+async function publish(request, response) {
+
+	if (response.ok) {
+		const blob = await response.blob();
+		const objectUrl = URL.createObjectURL(blob);
+		request.resolve(objectUrl);
+	} else {
+		request.reject(SingleFailedReason);
+	}
+}
+
+async function flushQueue() {
+
+	timer = 0;
+	state = StateFetching;
+
+	if (queue.length <= 0) {
+		state = StateIdle;
+		return;
+	}
+
+	const requests = queue;
+
+	queue = [];
+
+	const resources = requests.map(item => item.resource);
+	const bodyObject = { resources };
+	const bodyText = JSON.stringify(bodyObject);
+
+	const res = await fetch(documentLocaleSettings.oslo.batch, {
+		method: 'POST',
+		body: bodyText,
+		headers: { [ContentTypeHeader]: ContentTypeJson }
+	});
+
+	if (res.ok) {
+
+		const responses = (await res.json()).resources;
+
+		const tasks = [];
+
+		for (let i = 0; i < responses.length; ++i) {
+
+			const response = responses[i];
+			const request = requests[i];
+
+			const responseValue = new Response(response.body, {
+				status: response.status,
+				headers: response.headers
+			});
+
+			// New version might be available since the page loaded, so make a
+			// record of it.
+
+			const nextVersion = responseValue.headers.get(ETagHeader);
+			if (nextVersion) {
+				setVersion(nextVersion);
+			}
+
+			const cacheKey = new Request(formatCacheKey(request.resource));
+			const cacheValue = responseValue.clone();
+
+			if (cache === undefined) {
+				if (cachePromise === undefined) {
+					cachePromise = caches.open(CacheName);
+				}
+				cache = await cachePromise;
+			}
+
+			console.log(`[Oslo] cache prime: ${request.resource}`);
+			tasks.push(cache.put(cacheKey, cacheValue));
+			tasks.push(publish(request, responseValue));
+		}
+
+		await Promise.all(tasks);
+
+	} else {
+
+		for (const request of requests) {
+
+			request.reject(BatchFailedReason);
+		}
+	}
+
+	if (queue.length > 0) {
+		setTimeout(flushQueue, 0);
+	} else {
+		state = StateIdle;
+	}
+}
+
+function debounceQueue() {
+
+	if (state !== StateIdle) {
+		return;
+	}
+
+	if (timer > 0) {
+		clearTimeout(timer);
+	}
+
+	timer = setTimeout(flushQueue, DebounceTime);
+}
+
+function fetchWithQueuing(resource) {
+
+	const promise = new Promise((resolve, reject) => {
+
+		queue.push({ resource, resolve, reject });
+	});
+
+	debounceQueue();
+
+	return promise;
+}
+
+function formatCacheKey(resource) {
+
+	return documentLocaleSettings.oslo.collection + resource;
+}
+
+async function fetchWithCaching(resource) {
+
+	if (cache === undefined) {
+		if (cachePromise === undefined) {
+			cachePromise = caches.open(CacheName);
+		}
+		cache = await cachePromise;
+	}
+
+	const cacheKey = new Request(formatCacheKey(resource));
+	const cacheValue = await cache.match(cacheKey);
+	if (cacheValue === undefined) {
+		console.log(`[Oslo] cache miss: ${resource}`);
```

Comment:
Wonder if it's worth implementing some kind of debug mode so we can enable/disable these console logs at our leisure?

Reply:
I forgot to remove these before opening the PR. How would you suggest adding a debug mode? I like that better than just killing the lines.

Reply:
Hmm. You could make it part of the configuration it grabs off the document. Or just a global that gets set. Or an `options` object that gets passed into the entrypoint of the API.

Reply:
I settled on making it a special export that you can opt-into during development for now.